crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+use crate::entry::{Descriptor, Entry};
+use crate::sandboxed_tty_writer::SandboxedTTYWriter;
+use crate::wasi::wasi_snapshot_preview1::WasiSnapshotPreview1;
+use crate::wasi::{types, AsBytes, Errno, Result};
+use crate::WasiCtx;
+use crate::{clock, fd, path, poll};
+use log::{debug, error, trace};
+use std::cell::Ref;
+use std::convert::TryInto;
+use std::io::{self, Read, Seek, SeekFrom, Write};
+use std::ops::DerefMut;
+use wiggle_runtime::{GuestBorrows, GuestPtr};
+
+impl<'a> WasiSnapshotPreview1 for WasiCtx {
+    fn args_get<'b>(
+        &self,
+        mut argv: GuestPtr<'b, GuestPtr<'b, u8>>,
+        mut argv_buf: GuestPtr<'b, u8>,
+    ) -> Result<()> {
+        trace!("args_get(argv_ptr={:?}, argv_buf={:?})", argv, argv_buf);
+
+        let mut argv_buf_offset: types::Size = 0;
+        let mut bc = GuestBorrows::new();
+
+        for arg in &self.args {
+            let arg_bytes = arg.as_bytes_with_nul();
+            let elems = arg_bytes.len().try_into()?;
+            let as_arr = argv_buf.as_array(elems);
+            let raw_ptr = as_arr.as_raw(&mut bc)?;
+            unsafe { &mut *raw_ptr }.copy_from_slice(arg_bytes);
```

Comment:
FWIW stylistically I'd group sets of statements that are all "participating in the unsafe" together rather than trying to minimize the `unsafe` block to just one expression.

Reply:
Cool, I'm fine with that. I have to admit when porting the syscalls to `wiggle-runtime` I was following the `fd_pread` example from `wiggle`'s WASI testcase :-)